### DISCUSSION TITLE
How to use ErrorProne 2.33.0 with Gradle toolchains

### DISCUSSION THREAD

**Discussion by marcindabrowski** [tbroyer/gradle-errorprone-plugin#107] [General]
My default JDK is 17, but in my Gradle build I'm using [toolchains](https://docs.gradle.org/current/userguide/toolchains.html), and set it to Java 11.
My configuration looks like this:

```kotlin
java {
    toolchain {
        languageVersion = JavaLanguageVersion.of(11)
    }
}
```

Gradle toolchains seems to be preferred way for building projects with different Java versions.
Error prone plugin is not working with this configuration.
Is is possible to use toolchains with Java version lower than 17?
If yes, please provide an example.

**Comment by marcindabrowski** [comment]
I think I found the solution.
Set Java toolchain to 21:
```kotlin
java {
    toolchain {
        languageVersion = JavaLanguageVersion.of(21)
    }
}
```
and set `release` option of JVM compiler to 11:
```kotlin
tasks.withType<JavaCompile> {
    options.release.set(11)
}
```

I'm also using Groovy/Spock for tests and had to do this:
```kotlin
tasks.withType<GroovyCompile> {
    javaLauncher.set(javaToolchains.launcherFor {
        languageVersion.set(JavaLanguageVersion.of(11))
    })
}
```

**Reply by tbroyer** [reply]
Some nitpicking: don't use `withType<…> { … }`, use `withType<…>().configureEach { … }` instead to avoid eager configuration: https://docs.gradle.org/current/userguide/task_configuration_avoidance.html#eager_apis_to_avoid

Alternatively, you could configure a JDK 11 toolchain globally, and configure a JDK 17+ toolchain specifically for JavaCompile tasks:
```kotlin
java {
    toolchain {
        languageVersion = JavaLanguageVersion.of(11)
    }
}

tasks.withType<JavaCompile>().configureEach {
    // Use JDK >= 17 for Error Prone compatibility, but still target Java 11
    javaCompiler = javaToolchains.compilerFor { languageVersion = JavaLanguageVersion.of(21) }
    options.release = 11
}
```

**Reply by marcindabrowski** [reply]
Thanks with the tip with `configureEach`.

Well, the approach with using toolchain with higher Java in JavaCompile and setting release there to lower do not looks good.
And my proposal is to use the higher/latest java and set the `release` flag to lower. And this approach seems to somehow comply with article you linked.

**Comment by tbroyer** [comment]
What do you mean by "Error prone plugin is not working with this configuration"?

Note that by using a JDK 11 toolchain, you need to use a version of Error Prone that's compatible with JDK 11, so at most 2.31.0 (as 2.32.0 started requiring JDK 17).

BTW, [some would argue](https://jakewharton.com/gradle-toolchains-are-rarely-a-good-idea/) that the "preferred way" to target Java 11 is to use `--release 11` with any >=11 JDK (in Gradle terms, don't use toolchains, and configure `tasks.withType<JavaCompile>().configureEach { options.release = 11 }`)

**Reply by marcindabrowski** [reply]
My point was how to not stick to `2.31.0` and be able to run the latest when targeting Java 11.

Error prone is saying that the latest version can be used to analyze projects targeting Java 11 or earlier, but it has to be run using at least Java 17, and you can use `--release 11`.

And, the workable solution to use it with toolchains is as I wrote in my previous comment.

**Comment by marcindabrowski** [comment]
What about adding example with `toolchains` set to the newer Java and lower `release`?
I think that not only me, would like to use latest ErrorProne and will use `toolchains`.
WDYT?
